atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresGetCandidateCellsForSweeping.java

```diff
@@ -0,0 +1,326 @@
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.ConnectionSupplier;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.FullQuery;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.SqlConnectionSupplier;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.ranges.RangeBoundPredicates;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.sweep.DbKvsGetCandidateCellsForSweeping;
+import com.palantir.common.base.ClosableIterator;
+import com.palantir.common.base.ClosableIterators;
+import com.palantir.nexus.db.sql.AgnosticLightResultRow;
+import com.palantir.nexus.db.sql.AgnosticLightResultSet;
+
+import gnu.trove.list.TLongList;
+import gnu.trove.list.array.TLongArrayList;
+
+// Two considerations that influenced the implementation:
+//
+//   1. Window functions (OVER) on Postgres seem to be significantly slower than equivalent aggregates (GROUP BY).
+//
+//   2. Limiting an outer query doesn't seem to work well, e.g.:
+//          SELECT ... FROM (
+//              SELECT ... FROM my_table
+//              ...
+//              ORDER BY row_name, col_name, ts
+//              ...
+//          ) sub
+//          WHERE ...
+//          ORDER BY row_name, col_name, ts
+//          LIMIT 1000
+//      Postgres will often choose a full table scan for the inner SELECT query, even when an index
+//      range scan is sufficient (because of the limit on the outer query). Postgres also doesn't
+//      have hints like Oracle, so one can't force it to use a certain plan. A trick with specifying
+//      a very large limit for the inner query seems to work, hinting the optimizer that only
+//      the first few rows matter:
+//          SELECT ... FROM (
+//              SELECT ... FROM my_table
+//              ...
+//              ORDER BY row_name, col_name, ts
+//              ...
+//              LIMIT 100000000000000000000 -- a very large number that is guaranteed to exceed the table size
+//          ) sub
+//          WHERE ...
+//          ORDER BY row_name, col_name, ts
+//          LIMIT 1000
+//      However, this looks very hacky and fragile, so instead we resort to simply limiting the inner query
+//      (i.e., limiting the number of examined cells rather than the number of returned candidates).
+//
+// So, our SQL query does the following:
+//
+//   1. Grab a page of SQL rows, starting at some (row_name, col_name, ts).
+//   2. Number the rows, starting with 1.
+//   3. Group the rows by cell key, i.e. (row_name, col_name), and compute a few aggregates:
+//          - Minimum and maximum row number
+//          - Minimum and maximum timestamp
+//          - An array of all timestamps (using ARRAY_AGG)
+//   4. If the THOROUGH strategy is being used, do a self-join to check if the values are empty.
```

Comment:
when we do this, do we need to worry about empty values coming from transactions that didn't actually commit? (for example if a transaction did a delete but then called abort() instead of commit())